File: bindings/android/scion-http3-android/src/main/kotlin/com/anapaya/scion/http3/ScionHttp3Request.kt
// Copyright 2026 Anapaya Systems
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.anapaya.scion.http3

import java.net.URI
import java.net.URISyntaxException
import kotlin.time.Duration

/**
 * One request: where to send it, what to send, and the per-request limits that apply to it.
 *
 * Immutable, and safe to issue more than once. Build one with [Builder].
 */
public class ScionHttp3Request internal constructor(
    /** The absolute `https` URL. */
    public val url: String,
    /** The method, upper-case by convention but sent exactly as given. */
    public val method: String,
    /** The header section, which does not include the `content-type` a body implies. */
    public val headers: ScionHttp3Headers,
    /** The body, or null when the request sends none. */
    public val body: ScionHttp3RequestBody?,
    /** Addresses to use instead of resolving [url]'s host. Empty means resolve it. */
    public val targets: List<ScionAddress>,
    /** Overrides the client's request timeout for this request; null keeps it. */
    public val requestTimeoutMillis: Long?,
    /** Overrides the client's response-body limit for this request; null keeps it. */
    public val maxResponseBodyBytes: Long?,
) {
    /** A builder holding everything in this request, for deriving a modified copy. */
    public fun newBuilder(): Builder = Builder(this)

    override fun toString(): String = "$method $url"

    /**
     * Accumulates a request. Not thread-safe; build one per request.
     *
     * Everything a builder can judge for itself is rejected here rather than at request time: a URL
     * that is not absolute `https`, a method that is not a token, a header a header section cannot
     * carry. That keeps a caller's mistake an [IllegalArgumentException] at the call that made it,
     * and leaves [ScionHttp3Exception] for what actually goes wrong on the network.
     */
    public class Builder {
        private var url: String? = null
        private var method: String = "GET"
        private var headers = ScionHttp3Headers.Builder()
        private var body: ScionHttp3RequestBody? = null
        private val targets = mutableListOf<ScionAddress>()
        private var requestTimeoutMillis: Long? = null
        private var maxResponseBodyBytes: Long? = null

        public constructor()

        internal constructor(request: ScionHttp3Request) {
            url = request.url
            method = request.method
            headers = request.headers.newBuilder()
            body = request.body
            targets += request.targets
            requestTimeoutMillis = request.requestTimeoutMillis
            maxResponseBodyBytes = request.maxResponseBodyBytes
        }

        /**
         * The absolute URL to request.
         *
         * Must be `https`. This is an HTTP/3-only client with nothing to fall back to, so a
         * plaintext URL is a mistake rather than a downgrade.
         */
        public fun url(url: String): Builder {
            this.url = url
            return this
        }

        /** `GET`, dropping any body already set. */
        public fun get(): Builder = method("GET", null)

        /** `POST` with a body. */
        public fun post(body: ScionHttp3RequestBody): Builder = method("POST", body)

        /** `PUT` with a body. */
        public fun put(body: ScionHttp3RequestBody): Builder = method("PUT", body)

        /** `PATCH` with a body. */
        public fun patch(body: ScionHttp3RequestBody): Builder = method("PATCH", body)

        /** `DELETE`, with a body only if one is given. */
        @JvmOverloads
        public fun delete(body: ScionHttp3RequestBody? = null): Builder = method("DELETE", body)

        /** `HEAD`, dropping any body already set. */
        public fun head(): Builder = method("HEAD", null)

        /** Any other method, with a body only if the method takes one. */
        @JvmOverloads
        public fun method(
            method: String,
            body: ScionHttp3RequestBody? = null,
        ): Builder {
            this.method = validateMethod(method)
            this.body = body
            return this
        }

        /** Sets a header, replacing every line that already has this name. */
        public fun header(
            name: String,
            value: String,
        ): Builder {
            headers.set(name, value)
            return this
        }

        /** Appends a header line, keeping any line that already has this name. */
        public fun addHeader(
            name: String,
            value: String,
        ): Builder {
            headers.add(name, value)
            return this
        }

        /** Replaces the whole header section. */
        public fun headers(headers: ScionHttp3Headers): Builder {
            this.headers = headers.newBuilder()
            return this
        }

        /** Removes every header line with this name. */
        public fun removeHeader(name: String): Builder {
            headers.removeAll(name)
            return this
        }

        /** Sets or clears the body without changing the method. */
        public fun body(body: ScionHttp3RequestBody?): Builder {
            this.body = body
            return this
        }

        /**
         * Sends this request to [target] instead of resolving the URL's host.
         *
         * The escape hatch, for a host that has no SCION address records: a server on a local
         * PocketSCION topology, or one reached before its records exist. Call it more than once to
         * offer several addresses, which are then raced as if resolution had returned them all.
         *
         * The URL keeps its host and port, and both still determine `:authority` and the
         * certificate that is accepted, so this changes only *where* the request is sent.
         */
        public fun target(target: ScionAddress): Builder {
            targets += target
            return this
        }

        /** Replaces the targets, emptying them when given an empty list. */
        public fun targets(targets: List<ScionAddress>): Builder {
            this.targets.clear()
            this.targets += targets
            return this
        }

        /** Overrides the client's request timeout for this request. Must be positive. */
        public fun requestTimeout(timeout: Duration): Builder =
            requestTimeoutMillis(timeout.inWholeMilliseconds)

        /**
         * Overrides the client's request timeout, in milliseconds.
         *
         * The `Duration` overload is the one to use from Kotlin; this one exists because
         * `kotlin.time.Duration` is awkward to construct from Java.
         */
        public fun requestTimeoutMillis(millis: Long): Builder {
            require(millis > 0) { "a request timeout must be positive, not $millis ms" }
            requestTimeoutMillis = millis
            return this
        }

        /**
         * Overrides the client's limit on the response body, in bytes. Must be positive.
         *
         * This is the limit that bounds memory: the body is collected before the response is
         * returned, so a limit applied afterwards cannot un-spend it.
         */
        public fun maxResponseBody(bytes: Long): Builder {
            require(bytes > 0) { "a response body limit must be positive, not $bytes bytes" }
            maxResponseBodyBytes = bytes
            return this
        }

        /**
         * @throws IllegalArgumentException if no URL was set, or it is not an absolute `https` URL.
         */
        public fun build(): ScionHttp3Request {
            val url = requireNotNull(url) { "a request needs a url" }
            validateUrl(url)
            val body = body
            // On the built section rather than on the builder: a builder is reusable, and mutating it
            // here would carry this body's content-type into the next request built from it, even one
            // with another body or none.
            var headers = headers.build()
            if (body?.contentType != null) {
                headers = headers.newBuilder().addIfAbsent("content-type", body.contentType).build()
            }
            return ScionHttp3Request(
                url = url,
                method = method,
                headers = headers,
                body = body,
                targets = targets.toList(),
                requestTimeoutMillis = requestTimeoutMillis,
                maxResponseBodyBytes = maxResponseBodyBytes,
            )
        }

        private companion object {
            private const val METHOD_SYMBOLS = "!#$%&'*+-.^_`|~"

            private fun validateMethod(method: String): String {
                require(method.isNotEmpty()) { "a method cannot be empty" }
                method.forEach { c ->
                    require((c.isLetterOrDigit() && c.code < 0x80) || c in METHOD_SYMBOLS) {
                        "\"$method\" is not a method: expected a token, for example GET or POST"
                    }
                }
                return method
            }

            private fun validateUrl(url: String) {
                val uri =
                    try {
                        URI(url)
                    } catch (e: URISyntaxException) {
                        throw IllegalArgumentException(
                            "\"$url\" is not a valid URL: ${e.reason}",
                            e,
                        )
                    }
                require(uri.isAbsolute) {
                    "\"$url\" is not absolute: a request URL needs a scheme and a host"
                }
                require(uri.scheme.equals("https", ignoreCase = true)) {
                    "\"$url\" is not https. This client speaks HTTP/3 over SCION only and has " +
                        "nothing to fall back to, so a plaintext URL cannot be served."
                }
                require(!uri.host.isNullOrEmpty()) { "\"$url\" has no host" }
            }
        }
    }
}

File: bindings/apple/scion-http3-swift/Sources/ScionHTTP3/ScionHttp3Request.swift
// Copyright 2026 Anapaya Systems
import Foundation

/// One request: where to send it, what to send, and the per-request limits that apply to it.
///
/// It is safe to issue more than once and to change after a copy was sent. Everything a request can
/// get wrong is reported when it is sent as `ScionHttp3Error.invalidRequest`.
public struct ScionHttp3Request: Sendable, Equatable {
    /// An HTTP method: an RFC 9110 token, sent exactly as written.
    public struct Method: Sendable, Hashable, RawRepresentable, ExpressibleByStringLiteral,
        CustomStringConvertible
    {
        public let rawValue: String

        public init(rawValue: String) {
            self.rawValue = rawValue
        }

        public init(_ rawValue: String) {
            self.rawValue = rawValue
        }

        public init(stringLiteral value: String) {
            self.rawValue = value
        }

        public var description: String { rawValue }

        public static let get = Method("GET")
        public static let post = Method("POST")
        public static let put = Method("PUT")
        public static let patch = Method("PATCH")
        public static let delete = Method("DELETE")
        public static let head = Method("HEAD")

        func validate() throws {
            if rawValue.isEmpty {
                throw ScionHttp3Error.invalidRequest(detail: "a method cannot be empty")
            }
            if !rawValue.unicodeScalars.allSatisfy(ScionHttp3Headers.isToken) {
                throw ScionHttp3Error.invalidRequest(
                    detail: "\"\(rawValue)\" is not a method: expected a token, "
                        + "for example GET or POST")
            }
        }
    }

    /// The absolute URL to request.
    ///
    /// Must be `https`. This is an HTTP/3-only client with nothing to fall back to.
    public var url: String

    /// The method. If not set, defaults to `GET`.
    public var method: Method

    /// The header section. It does not include the `content-type` the body implies; that line is
    /// added when the request is sent, unless the section already has one.
    public var headers: ScionHttp3Headers

    /// The body or nil when the request sends none.
    public var body: ScionHttp3RequestBody?

    /// Addresses to send the request to instead of resolving the URL's host. Empty means resolve.
    ///
    /// Several addresses are raced as if resolution had returned them all.
    public var targets: [ScionAddress]

    /// Overrides the client's request timeout for this request.
    public var requestTimeout: TimeInterval?

    /// Overrides the client's response-body limit for this request.
    public var maxResponseBodyBytes: Int?

    /// Sends this request to one address instead of resolving the URL's host.
    public var target: ScionAddress? {
        get { targets.first }
        set { targets = newValue.map { [$0] } ?? [] }
    }

    public init(url: String, method: Method = .get, body: ScionHttp3RequestBody? = nil) {
        self.url = url
        self.method = method
        self.headers = ScionHttp3Headers()
        self.body = body
        self.targets = []
    }

    /// Throws `ScionHttp3Error.invalidRequest` for the first thing the stack cannot be given.
    func validate() throws {
        try Self.validateUrl(url)
        try method.validate()
        try headers.validate()
        if let requestTimeout, !(requestTimeout > 0) || !requestTimeout.isFinite {
            throw ScionHttp3Error.invalidRequest(
                detail: "a request timeout must be positive and finite, not \(requestTimeout) s")
        }
        if let maxResponseBodyBytes, maxResponseBodyBytes <= 0 {
            throw ScionHttp3Error.invalidRequest(
                detail: "a response body limit must be positive, not \(maxResponseBodyBytes) bytes")
        }
    }

    private static func validateUrl(_ url: String) throws {
        guard let components = URLComponents(string: url) else {
            throw ScionHttp3Error.invalidRequest(detail: "\"\(url)\" is not a valid URL")
        }
        guard let scheme = components.scheme else {
            throw ScionHttp3Error.invalidRequest(
                detail: "\"\(url)\" is not absolute: a request URL needs a scheme and a host")
        }
        if scheme.lowercased() != "https" {
            throw ScionHttp3Error.invalidRequest(
                detail: "\"\(url)\" is not https. This client speaks HTTP/3 over SCION only and "
                    + "has nothing to fall back to, so a plaintext URL cannot be served.")
        }
        if components.host?.isEmpty ?? true {
            throw ScionHttp3Error.invalidRequest(detail: "\"\(url)\" has no host")
        }
    }
}

extension ScionHttp3Request: CustomStringConvertible {
    public var description: String { "\(method) \(url)" }
}
